### ISSUE TITLE
no-unused-vars is not reporting for object destructuring

### ISSUE THREAD

**Issue by cassus** [babel/babel-eslint#142]
No error:
```js
const {connect} = require('redux/react')
```

Error
`error  no-unused-vars  connect is defined but never used`
```
const connect = require('redux/react').connect
```

**Comment by sebmck** [comment]
Please verify that this doesn't occur in default ESLint.

On Sunday, 5 July 2015, Adam Banko <notifications@github.com> wrote:

> No error:
>
> const {connect} = require('redux/react')
>
> Error
> error no-unused-vars connect is defined but never used
>
> const connect = require('redux/react').connect
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel-eslint/issues/142>.
>


-- 
Sebastian McKenzie

**Comment by cassus** [comment]
Confirmed.

app/utils.js :
```
const {Bacona} = require('baconjs')
```
---
` //"parser": "babel-eslint",`
```
$ ./node_modules/.bin/eslint app/utils.js 

app/utils.js
  2:7  error  Bacona is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

---
`"parser": "babel-eslint",`
```
$ ./node_modules/.bin/eslint app/utils.js 
$ 
```

**Comment by hzoo** [comment]
Ah it's that it's not reporting it. Ok due to https://github.com/babel/babel-eslint/pull/122 - will figure out a fix for 3.1.21.

#122 was for `var {field1, field2, ...clone} = originalObject;` - allowing field1, and field2 to be unused. So I guess we would just ignore the fields left of the spread (if there is one)? Looks like I wrongly visited all the properties instead.

**Comment by deepsweet** [comment]
v3.1.21 gives me a lot of `no-unused-vars` errors because of this case:

```js
test(data) {
    return {
        foo: 'bar',
        ...data
    };
}
```

but with v3.1.20 everything is ok.

**Comment by hzoo** [comment]
@sebmck I guess it did affect `no-unused-vars`.. What do you think about introducing another property to track if it's a `SpreadProperty`?

```js
 if (this.isSpreadProperty()) {
  // to track if it's a spread property
  node._spread = true;
  ...
}
```

**Comment by sebmck** [comment]
Yeah, a "private" variable would be a good solution.

**Comment by hzoo** [comment]
@deepsweet should be fixed in 3.1.22

**Comment by jquense** [comment]
ooo i'd like to make use of this in: https://github.com/babel/eslint-plugin-babel/blob/master/rules/object-shorthand.js#L31 instead of the regex, how do you feel about the plugin using these private properties?

**Comment by sebmck** [comment]
Actually, maybe something like `node._babelType = node.type;` at the top of the AST normaliser visitor. Happy for that to a pseudo-public API.

**Comment by jquense** [comment]
That would be great, seems like a good general solution and removes my need to sometimes try and double parse the source again or sniff with regexs, both being fraught.

**Comment by rui-infotrack** [comment]
```
error no-unused-vars connect is defined but never used

const connect = require('redux/react').connect
```

This error seems to come back in 4.0.5:

```javascript
import { connect } from 'redux/react';

@connect(state => ({
  hello: state.world
}))
```

**Comment by zallek** [comment]
got the same problem

**Comment by hzoo** [comment]
Looks like an issue with decorators rather than the original issue?
And it might be a babel issue which should be fixed in the next release of babel - if not then make a new issue for this please.
